Data-Structures/Ruby/binary-tree.rb

```diff
@@ -0,0 +1,70 @@
+class Node
+  attr_accessor :parent, :left, :right, :data
+  def initialize(data = nil, parent = nil, left = nil, right = nil)
+    @parent, @left, @right, @data = parent, left, right, data
+  end
+end
+
+class BinaryTree
+  attr_accessor :length
+
+  def insert data
+    return create_root(data) if @root.nil?
+    new_node = Node.new(data)
+    parent = @leaf_nodes.first
+
+    if parent.left.nil?
+      parent.left = new_node
+    elsif parent.right.nil?
+      parent.right = new_node
+      @leaf_nodes.shift
+    end
+
+    @leaf_nodes << new_node
+    @length += 1
+  end
+
+  def delete data
+    dfs(@root).each do |node|
+      if node.data == data
+        @length -= 1
+        node.data = nil
+      end
+    end
+  end
+
+  def traverse_dfs &block
+    dfs(@root).map(&:data).compact.each { |data| block.call(data) }
+  end
+
+  def traverse_bfs &block
+    bfs.map(&:data).compact.each { |data| block.call(data) }
```

Comment:
This is clever, but it seems computationally inefficient. BFS already *does* traverse each node so why not call the block directly in that?